src/ray/gcs/gcs_server/gcs_node_manager.cc

```diff
@@ -335,6 +337,77 @@ void GcsNodeManager::HandleGetAllAvailableResources(
   GCS_RPC_SEND_REPLY(send_reply_callback, reply, Status::OK());
 }
 
+void GcsNodeManager::HandleGetAllHeartbeat(const rpc::GetAllHeartbeatRequest &request,
+                                           rpc::GetAllHeartbeatReply *reply,
+                                           rpc::SendReplyCallback send_reply_callback) {
+  if (!node_heartbeats_.empty()) {
+    auto batch = std::make_shared<rpc::HeartbeatBatchTableData>();
+    absl::flat_hash_map<ResourceSet, rpc::ResourceDemand> aggregate_load;
+    for (auto &heartbeat : node_heartbeats_) {
+      // Aggregate the load reported by each raylet.
+      auto load = heartbeat.second.resource_load_by_shape();
+      for (const auto &demand : load.resource_demands()) {
+        auto scheduling_key = ResourceSet(MapFromProtobuf(demand.shape()));
+        auto &aggregate_demand = aggregate_load[scheduling_key];
+        aggregate_demand.set_num_ready_requests_queued(
+            aggregate_demand.num_ready_requests_queued() +
+            demand.num_ready_requests_queued());
+        aggregate_demand.set_num_infeasible_requests_queued(
+            aggregate_demand.num_infeasible_requests_queued() +
+            demand.num_infeasible_requests_queued());
+        if (RayConfig::instance().report_worker_backlog()) {
+          aggregate_demand.set_backlog_size(aggregate_demand.backlog_size() +
+                                            demand.backlog_size());
+        }
+      }
+      heartbeat.second.clear_resource_load_by_shape();
+
+      batch->add_batch()->Swap(&heartbeat.second);
+    }
+
+    for (auto &demand : aggregate_load) {
```

Comment:
Can we take these logic out from the SendBatchedHeartbeat? I think it won't have any impact because the resource demand and placement group load is only used for the autoscaler. Please lmk if I am wrong @ericl.

Reply:
Oh nvm. It looks like it is already handled.

Reply:
This is a nice catch!
We have removed that before in this pr but it came back after merging the master. Only can we do is remove it again :(